src/main/kotlin/org/rust/lang/core/lexer/RustLexerUtil.kt

```diff
@@ -1,17 +1,28 @@
 package org.rust.lang.core.lexer
 
+import com.intellij.lexer.Lexer
+
+fun CharSequence.tokenize(lexer: Lexer): Sequence<Lexer> =
+    generateSequence({
+        lexer.start(this)
+        lexer.tokenType?.let { lexer }
+    }, { currentLexer ->
+        currentLexer.advance()
+        currentLexer.tokenType?.let { currentLexer }
+    })
```

Comment:
Hm, this does not look very nice to me. You can accidentally call `advance` on the returned lexer.

What about introducing `data class Lexer.Token(val start: Int, val end: Ind, val type: IElementType)`? Or maybe this all is overengineered, and we should stick to imperative API and go back to `while` loops?